### PR TITLE
Remove libpagure

### LINKED ISSUE #64: libpagure has changed
libpagure==0.20 changed `headers` handling, so that
code like:
https://github.com/packit-service/ogr/blob/master/ogr/services/our_pagure.py#L172
has to be adapted, currenly I've added dependency on lower version of libpagure to 
https://github.com/packit-service/ogr/pull/63/files#diff-380c6a8ebbbce17d55d50ef17d3cf906R44
to avoid this issue.

### LINKED ISSUE #67: libpagure.Pagure expects '/' not in repo attribute
While debugging https://github.com/packit-service/packit/issues/305 I realized that it fails because [Pagure.create_basic_url()](https://pagure.io/libpagure/blob/17e240ea337e6962aa660b0cfd174ef6309f4162/f/libpagure/libpagure.py#_115) creates:
`https://src.fedoraproject.org//api/0/rpms/rpms/packit/pull-requests` which does not exist, correct is:
`https://src.fedoraproject.org//api/0/rpms/packit/pull-requests`.

The problem is that [OurPagure.repo](https://pagure.io/libpagure/blob/17e240ea337e6962aa660b0cfd174ef6309f4162/f/libpagure/libpagure.py#_115) attribute has a value `rpm/packit`, while libpagure probably expects it to be `packit` only.

Anybody has any idea whether this is a `ogr` or `packit` or `libpagure` issue?
I tend to think it's `ogr` issue, because [OurPagure.repo_name](https://github.com/packit-service/ogr/blob/master/ogr/services/our_pagure.py#L22) expects `/` in `self.repo`, which is probably not what `libpagure` wants.

And it gets even more puzzled when I tell you that [Pagure.create_basic_url() has changed](https://pagure.io/libpagure/c/075ee2f17c36b1c1477f72f711204945b609d5b1?branch=master) even more recently.

### PR DESCRIPTION
**Pull request by lachmanfrantisek**
- Remove OurPagure.
- Implement forking logic and PR methods.
- First part of implementing the Pagure classes without libpagure.
- Add string representation for PRComment and PullRequest.

Fix #64
Fix #67

---

TODO:

- [x] readonly mode
- [x] tests

---

## How to test?

```python
from ogr.services.pagure import PagureService
from ogr.abstract import PRStatus


service = PagureService(token="???_PAGURE_TOKEN_???")

# Play with the API

print(service.user.get_username())
project = service.get_project(repo="colin", namespace="rpms")
print(project.get_commit_statuses("339a19b0bbc766d0c6cdbbc2ef5a32c0de9f7551")[0])
print(project.get_tags())

fork = project.get_fork()
print(fork.exists())

project.fork_create()

for pr in project.get_pr_list(status=PRStatus.all):
    print(pr)

pr = project.get_pr_info(3)
print(pr)
for c in project.get_pr_comments(3):
    print(c)

print(fork.is_fork)
print(fork.is_forked())
```

### CODE REVIEW

**Comment by lachmanfrantisek** [comment]
@jscotka 
I've removed the old read-only setup for libpagure. Do you have some idea how it can be done in the current state without libpagure?
Thanks!

**Comment by rpitonak** [comment]
I played with it a little bit, and I got:
```
>>> project = service.get_project(repo="conu", namespace="rpms")
>>> for pr in project.get_pr_list(status=PRStatus.all):
...     print(pr)
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/home/vagrant/ogr/ogr/abstract.py", line 39, in __str__
    f"PullRequest("
TypeError: 'NoneType' object is not subscriptable
```

**Comment by lachmanfrantisek** [comment]
@rpitonak 

Good catch. Fixed.

**Comment by rpitonak** [comment]
> Good catch. Fixed.

Thanks, I tried all of the commands above on `conu` and it worked smoothly. Nice work so far!

**Comment by lachmanfrantisek** [comment]
> Thanks, I tried all of the commands above on conu and it worked smoothly.

That was just an example. Whole API should be already implemented...

File: ogr/abstract.py [hunk 1]
```diff
@@ -48,12 +65,44 @@ def __init__(
...
+        return (
+            f"PRComment("
+            f"comment='{comment}', "
+            f"author='{self.author}', "
+            f"created='{self.created}', "
+            f"edited='{self.edited}')"
+        )
+
 
 class CommitStatus:
```
Comment (lachmanfrantisek) [inline ogr/abstract.py]: I do not want to break backwards compatibility, but what about using `flag` for this thing?

File: ogr/abstract.py [hunk 2]
```diff
@@ -89,6 +138,15 @@ def save_archive(self, filename):
         file.close()
 
 
+class GitTag:
+    def __init__(self, name: str, commit_sha: str) -> None:
+        self.name = name
+        self.commit_sha = commit_sha
+
+    def __str__(self) -> str:
+        return f"GitTag(name={self.name}, commit_sha={self.commit_sha})"
```
Comment (lachmanfrantisek) [inline ogr/abstract.py]: TODO: Add a link to `GitTag` from `Release`.

File: ogr/services/pagure.py [hunk 3]
```diff
@@ -147,51 +247,80 @@ def pr_comment(
...
+        if filename is not None:
+            payload["filename"] = filename
+        if row is not None:
+            payload["row"] = row
+
+        self._call_project_api(
+            "pull-request", pr_id, "comment", method="POST", data=payload
         )
 
+        return PRComment(comment=body, author=self.service.user.get_username())
```
[the remark this answers is not in the document]
Comment (lachmanfrantisek) [reply ogr/services/pagure.py]: We do not know the exact time and do not get anything useful as a response from the API.

What about calling the API ones again and try to search that comment?

File: ogr/abstract.py (hunk 1 above)
Comment (TomasTomecek) [reply ogr/abstract.py]: what flag?

File: ogr/services/pagure.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TomasTomecek) [reply ogr/services/pagure.py]: can we do that with a single API call to get all comments on a PR?

File: ogr/services/pagure.py [hunk 4]
```diff
@@ -24,113 +27,210 @@ def __init__(
...
+
+        if raw:
+            return response
 
-        Only for this instance and newly created Projects via get_project.
+        try:
+            output = response.json()
+        except Exception as err:
+            logger.debug(response.text)
+            raise PagureAPIException("Error while decoding JSON: {0}".format(err))
```
[the remark this answers is not in the document]
Comment (lachmanfrantisek) [reply ogr/services/pagure.py]: Is that ok?

File: ogr/abstract.py (hunk 1 above)
Comment (lachmanfrantisek) [reply ogr/abstract.py]: `CommitFlag`/`PullRequestFlag`...

File: ogr/services/pagure.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lachmanfrantisek) [reply ogr/services/pagure.py]: - one API call for the comment creation
- one API call for getting all comments (but we need to find the one we created)

[the remark this answers is not in the document]
Comment (lachmanfrantisek) [reply ogr/services/pagure.py]: I will go with the partial response for now. We can use lazy-evaluation/second-call later. Is that ok?

File: ogr/services/pagure.py [hunk 5]
```diff
@@ -24,113 +27,224 @@ def __init__(
...
+        It returns the raw JSON returned by the API or raises an exception
+        if something goes wrong.
+
+        It can return raw response when raw=True.
         """
-        Change an API token.
+        method = method or "GET"
+
+        try:
+            response = self.session.request(
```
Comment (jscotka) [inline ogr/services/pagure.py]: This should use ``self.get_raw_request`` instead of duplication same code here.

File: ogr/services/pagure.py [hunk 6]
```diff
@@ -24,113 +27,224 @@ def __init__(
...
+
+        req = self.session.request(
+            method=method,
+            url=url,
+            params=params,
+            headers=self.header,
+            data=data,
+            verify=not self.insecure,
+        )
+        return req
```
Comment (jscotka) [inline ogr/services/pagure.py]: please inspire this code by: https://github.com/PyGithub/PyGithub/blob/master/github/Requester.py#L367
especially return of this method: ``return status, responseHeaders, output`` what then could be easily used for mocking

**Comment by jscotka** [comment]
> @jscotka
> I've removed the old read-only setup for libpagure. Do you have some idea how it can be done in the current state without libpagure?
> Thanks!

@lachmanfrantisek by readonly mode are directly wrappend method in services/pagure, so no change here.

**Comment by lachmanfrantisek** [comment]
> @lachmanfrantisek by readonly mode are directly wrappend method in services/pagure, so no change here.

Sorry, I meant the request caching...

File: ogr/abstract.py (hunk 1 above)
Comment (TomasTomecek) [reply ogr/abstract.py]: Sounds good. You can even keep this class, mark it as deprecated and use inheritance to introduce the Flag classes.

**Comment by lachmanfrantisek** [comment]
@jscotka Can you take another look?

**Comment by lachmanfrantisek** [comment]
> Hi Frantisek, now it seems very well from perspective of mocking and representing request-response object. thanks a lot for these changes

Thanks, but the loading response values does not work now...

**Comment by lachmanfrantisek** [comment]
@jscotka Do you know, why the loading does not work and I need to dump to dict manually (c31b895)?

**Comment by lachmanfrantisek** [comment]
I have a working version with direct yaml dumping but it does not go through the pre-commit yaml validation.

@jscotka Do we want to use `unsafe` in pre-commit validation or use current setup?

**Comment by lachmanfrantisek** [comment]
OK, let's leave it as is and we can change it to object-dumping later.

**Comment by lachmanfrantisek** [comment]
@TomasTomecek @jpopelka Do you want to take a look at this? Otherwise, we can merge.

File: ogr/abstract.py (hunk 1 above)
Comment (lachmanfrantisek) [reply ogr/abstract.py]: I've created a new issue so we can merge this.

https://github.com/packit-service/ogr/issues/73

File: ogr/abstract.py (hunk 2 above)
Comment (lachmanfrantisek) [reply ogr/abstract.py]: I've created a new issue for this so we can merge: https://github.com/packit-service/ogr/issues/74

**Comment by jpopelka** [comment]
Please merge if it's ready, so we can release a new version.

**Comment by TomasTomecek** [comment]
Let's merge and test in production :P

**Comment by jpopelka** [comment]
> test in production

Do we need to do any changes in packit? I still see `OurPagure` in [packit's tests](https://github.com/packit-service/packit/blob/master/tests/conftest.py#L37)

**Comment by lachmanfrantisek** [comment]
> Do we need to do any changes in packit? I still see OurPagure in packit's tests.

Good point.

The API should state the same, but mocking probably needs some changes...

**Comment by jpopelka** [comment]
Filled new issue https://github.com/packit-service/packit/issues/355

**Pull request merged**